subsys/mgmt/hawkbit/hawkbit.c

```diff
@@ -1197,240 +1103,468 @@ static bool check_hawkbit_server(void)
-			 HAWKBIT_STATUS_EXEC_NONE)) {
+	ret = hawkbit_parse_deployment(&s->hb_context, &s->hb_context.results.dep, url_buffer,
+				       NULL);
+	if (ret < 0) {
+		LOG_ERR("Failed to parse deploymentBase: %d", ret);
+		s->hb_context.code_status = HAWKBIT_METADATA_ERROR;
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+		return;
+	}
+
+	flash_img_init(&s->hb_context.flash_ctx);
+
+	if (!send_request(&s->hb_context, HAWKBIT_DOWNLOAD, url_buffer, NULL)) {
 		LOG_ERR("Send request failed (%s)", "HAWKBIT_DOWNLOAD");
-		hb_context.code_status = HAWKBIT_NETWORKING_ERROR;
-		goto cleanup;
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+		return;
 	}
 
-	if (hb_context.code_status == HAWKBIT_DOWNLOAD_ERROR) {
-		goto cleanup;
+	if (s->hb_context.code_status == HAWKBIT_DOWNLOAD_ERROR) {
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+		return;
 	}
 
 	/* Check if download finished */
-	if (!hb_context.final_data_received) {
+	if (!s->hb_context.final_data_received) {
 		LOG_ERR("Download incomplete");
-		hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
-		goto cleanup;
+		s->hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+		return;
 	}
 
 	/* Verify the hash of the stored firmware */
-	fic.match = hb_context.dl.file_hash;
-	fic.clen = hb_context.dl.downloaded_size;
-	if (flash_img_check(&hb_context.flash_ctx, &fic, FIXED_PARTITION_ID(SLOT1_LABEL))) {
+	struct flash_img_check fic = {0};
+
+	fic.match = s->hb_context.dl.file_hash;
+	fic.clen = s->hb_context.dl.downloaded_size;
+	if (flash_img_check(&s->hb_context.flash_ctx, &fic, FIXED_PARTITION_ID(SLOT1_LABEL))) {
 		LOG_ERR("Failed to validate stored firmware");
-		hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
-		goto cleanup;
+		s->hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+		return;
 	}
 
 	/* Request mcuboot to upgrade */
 	if (boot_request_upgrade(BOOT_UPGRADE_TEST)) {
 		LOG_ERR("Failed to mark the image in slot 1 as pending");
-		hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
-		goto cleanup;
+		s->hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
+		smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+		return;
 	}
 
 	/* If everything is successful */
-	hb_context.code_status = HAWKBIT_UPDATE_INSTALLED;
-	hawkbit_device_acid_update(hb_context.json_action_id);
+	s->hb_context.code_status = HAWKBIT_UPDATE_INSTALLED;
+	hawkbit_device_acid_update(s->hb_context.json_action_id);
 
-	hb_context.dl.http_content_size = 0;
+	smf_set_state(SMF_CTX(s), &hawkbit_states[S_HAWKBIT_TERMINATE]);
+}
 
-cleanup:
-	cleanup_connection();
+static void s_terminate(void *o)
+{
+	struct s_object *s = (struct s_object *)o;
 
-error:
-	free(hb_context.response_data);
-	k_sem_give(&probe_sem);
-	return hb_context.code_status;
+	smf_set_terminate(SMF_CTX(s), s->hb_context.code_status);
```

Comment:
Can this ever be zero? e.g. with a successful download?

Reply:
No

Reply:
> Yes it does, but not including today's changes.

Heh. okay.